semgrep/semgrep/config_resolver.py

```diff
@@ -133,7 +133,16 @@ def get_rules(self, no_rewrite_rule_ids: bool) -> List[Rule]:
             # re-write the configs to have the hierarchical rule ids
             configs = self._rename_rule_ids(configs)
 
-        return [rule for rules in configs.values() for rule in rules]
+        rules = []
+        seen = set()
+        for rule in [rule for rules in configs.values() for rule in rules]:
+            if rule.id in seen:
```

Comment:
Rule IDs are not a one-to-one mapping with a rule ([see e.g. here](https://github.com/returntocorp/semgrep-rules/blob/develop/go/lang/correctness/permissions/file_permission.yaml)), so we can't use `rule.id` to filter seen rules. If we define a reasonable `__eq__` comparator on the `rule.Rule` class, then we could simply do something like the following here:

```
return set(rule for rules in configs.values() for rule in rules)
```

And the `set` will handle everything for us :+1:

Reply:
Hi @mschwager!

Good to know; I will have to extend Rule with `__eq__` and `__hash__`.

Unfortunately, we can't use sets here for the rules themselves; the order being non-deterministic causes serious problems with the snapshot tests.  I talked to @brendongo about this on Slack, and the best option that we came to was to continue maintaining them as a list for the moment pending a more significant overhaul of this section of the code.

I experimented with passing it through `dict.fromkeys`, however, with Python 3.6, we'd be relying on a Cython implementation detail; dict was not guaranteed to preserve insertion order until Python 3.7.

The other option is to pass it through set as an intermediary and then sort the output, however, that would require adding `__lt__` and regenerating all the tests.

Happy to do either; let me know which is preferable.

Reply:
Hey there,

Ahh, interesting, good catch. We could use `OrderedDict.fromkeys` - something like:

```python
# Remove duplicates, preserve order
return list(OrderedDict.fromkeys(rule for rules in configs.values() for rule in rules))
```

Would that minimize snapshot thrashing?